### ISSUE TITLE
Expanded -short, original-, reviewed- variables

### ISSUE THREAD

**Issue by bwiernik** [citation-style-language/schema#167]
Recently, in response to user needs for the SBL style, citeproc-js added support for providing `-short` versions of all CSL variables, to be rendered with `form="short"`. (https://forums.zotero.org/discussion/comment/324592/#Comment_324592)

In my work on apa.csl, I'm finding that it wants a lot more detailed information for reviews (e.g., medium of item being reviewed, date of item being reviewed), as well as for original publication information (e.g., original medium, original container title, original pages, original editor) than is currently possible with existing CSL variables. As far as I am aware, MLA and Chicago have similar requirements.

I suggest that `-short`, `original-`, `reviewed-` should be expanded so that they can be applied to any CSL variable. This would allow maximum flexibility without having to individually specify each possible variable of this kind.

**Comment by denismaier** [comment]
Sounds good to me. Any potential drawbacks when this is implemented?

**Comment by bwiernik** [comment]
These could be implemented in CSL-JSON as arrays for `short`, `original`, `reviewed`: https://github.com/citation-style-language/schema/issues/169

**Comment by denismaier** [comment]
Hmm. @dstillman Looks Zotero devs are not big fans arrays and objects. Suggestions concerning the data structure here?

**Comment by dstillman** [comment]
> These could be implemented in CSL-JSON as arrays for short, original, reviewed.

I'm not clear what the suggestion is here. Can you give an example?

**Comment by bwiernik** [comment]
I should have said objects, not arrays. Example:

```
"reviewed": {
  "type": "motion_picture",
  "medium": "DVD",
  "title": "Title of reviewed movie"
}
```

vs listing these as individual `reviewed-` variables.
```
reviewed-type: motion_picture
reviewed-medium: DVD
reviewed-title: Title of reviewed movie
```

**Comment by bdarcus** [comment]
So your wanting to change a mostly flat (aside from contributors and dates) data model to a more structured one.

**Comment by denismaier** [comment]
Ultimately, this could lead to a data model as outlined [here](https://github.com/citation-style-language/csl-evolution/issues/12#issuecomment-617029729).

**Comment by dstillman** [comment]
I think this is more a discussion for the CSL list, but in general I would strongly advocate for key-value pairs over objects, _except_ where the fields don't make sense independently and the app would need special handling of all associated variables for proper processing anyway. If it's something where there could be a direct mapping between a field and a variable, it's vastly simpler to stick to key-value pairs, and it also allows for hacks like Extra. Reducing implementation complexity is much more important in my view than reducing verbosity in CSL-JSON.

**Comment by bdarcus** [comment]
Yeah, it's easy to add these variable strings ("reviewed-ttile" and
such), so let's just do that. We could have defined "container" as an
object, for example, but we didn't.

**Comment by denismaier** [comment]
I understand.
But what about special handling if prefixes and affixes to variables? Is there a way to define affixes that could be used on other variables? Like allow -short as a general modifing suffix and reviewed- as general modifing prefix? Would that be somehow possible?

**Comment by dstillman** [comment]
I think that would affect the processor more than the app. If we support a given `-short` or `reviewed-` field, the mapping would be hard-coded. It's the processor that would need to know how to handle those.

(I don't totally get it, though. Wouldn't there be nonsensical possibilities? What does `issued-short` mean?)

**Comment by bwiernik** [comment]
Okay, so let's stick with key-value pairs.

Dan makes a good point on `-short`. It should apply only to standard variables (string, number, title), not name or date variables.

**Comment by denismaier** [comment]
So 2 questions:

1. Is it possible to have such prefix/suffix rules in JSON to prevent unnecessary verbosity? (and yes, we will need to restrict `-short` to certain variables)

2. If yes, should we do this? 

Or should we simply add possible variables to `reviewed-`, `original-`, `container-`, `collection-` ?

**Comment by bwiernik** [comment]
The three relevant affixes are `-short`, `original-` and `reviewed-`. Could we define valid combinations of these in the style and data schemas using string concatenation?

So, something like `variables.short = variables.standard + '-short'`?

**Comment by denismaier** [comment]
> The three relevant affixes are `-short`, `original-` and `reviewed-`.

Addendum: With `container-`, `collection-` I was suggesting we should add that now. But perhaps in the medium run?

> So, something like `variables.short = variables.standard + '-short'` and `variables.original = 'original-' + variables.all`?

That looks good. Would make schema updates easier, wouldn't it? (But I'm a bit pessimistic that will work so easily: https://stackoverflow.com/questions/9708192/use-a-concatenated-dynamic-string-as-javascript-object-key yes that's old, but I perhaps still relevant? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer says it's possible with recent JS, but not with JSON.)

**Comment by bwiernik** [comment]
I think container and collection are a much bigger can of worms than the others, so let's set those aside for now.

**Comment by bwiernik** [comment]
For the data schema, an option would be to split out the schemas into separate files that match the RNC type and variable structure and use a build script to compile them at commit time.

**Comment by denismaier** [comment]
How would that look like and how would that solve that problem with schema verbosity?

**Comment by denismaier** [comment]
Idea: if we have all variables available with `original-` or `reviewed-`, what about a mechanism like `alternative` in csl-m where you can render all variables prefixed with `alt-` with a single `alternative` variable. Could make style coding easier.

**Comment by bwiernik** [comment]
In addition to or instead of making them available as regular variables? It would need to be in addition to if anything.

**Comment by bwiernik** [comment]
> How would that look like and how would that solve the problem with schema verbosity? Would it?

A python script in GitHub Actions could compile the csl-data.json at commit time. It would have a list of all types and variables (separated by category) and dynamically construct the JSON. The main benefit would be ease of maintenance and updating, not needing to manually keep four nearly identical lists aligned manually.

**Comment by bdarcus** [comment]
> For the data schema, an option might be to split out the schemas into separate files that match the RNC type and variable structure and use a build script to compile them at commit time.

Yes, I was wondering about something like this.

**Comment by denismaier** [comment]
> In addition to or instead of making them available as regular variables? It would need to be in addition to if anything.

Sure, in addition to the regular variables. For reviews you will most likely want a full reference, right? (And that reference should also be rendered according to the current style---so giving these details in the regular title is actually not ideal.)

**Comment by denismaier** [comment]
So, what shall we do about this now? Should I draft a PR for `original-`, `reviewed-`, and `-short`? Or should we go the automated route instead?

**Comment by bdarcus** [comment]
> So, what shall we do about this now? Should I draft a PR for `original-`, `reviewed-`, and `-short`? Or should we go the automated route instead?

Depends who's going to write the python script and when.

I have basic python skills, but am not knowledge about parsing text as we need (see [comment](https://github.com/citation-style-language/schema/issues/223#issuecomment-644051259)).

**Comment by denismaier** [comment]
> I have basic python skills, but am not knowledge about parsing text as we need

The question is: How will our input look like? Will we just use the json? Or could we even work with native python structures? If so, we don't have to parse anything.

**Comment by bdarcus** [comment]
I don't understand. I was assuming input is the rnc file(s), output is
csl-data.json.

What were you thinking?

**Comment by denismaier** [comment]
A was thinking we could use a common source for both rnc and json.

```python

variables = [
    {
        "name" : "title",
        "type" : "string",
        "variants" : ["original-", "reviewed-", "-short"]
    },
    {
        "name" : "author",
        "type" : "name",
        "variants" : ["original-", "reviewed-", "container-]
    },
    ]

def create_rnc(variables):
    # this creates the rnc schema variable list
    return rnc

def create_json(variables):
    # this creates the json schema
    return json

rnc= create_rnc()
json = create_json()

```

**Comment by denismaier** [comment]
> A single, say python, file, whose contents is the data representation, output to both rnc and json?

Exactly, see above. (Or instead of python, we could also use some other common source that is easy to write and parse, say yaml or toml.

**Comment by bdarcus** [comment]
IC.

I'm agnostic; whatever gets us to the best and easiest result, which is
consistent schemas, and clean git histories, including diffs.

I'm not sure on the details of CI in GitHub; how it would work.

On your example, though, maybe better to have separate dicts for datatypes;
like "variables-string."

**Comment by bdarcus** [comment]
You want to test a minimal toml or yaml approach in your test repo and
report?

;-)

**Comment by bdarcus** [comment]
PS - let's move this conversation to the linked issue?

**Comment by denismaier** [comment]
Maybe we're making things too complicated regarding the json here?

Can't we just add a note to the specs that would basically say, "Any non-number variable can also be supplied in the short form with the suffix -short. For every variables exist variants prefixed with original- and reviewed-".
I mean, the json is not used for validation, right? Do we need to have those in the json schema at all?

**Comment by denismaier** [comment]
And regarding the rnc, is there any way to use patterns without using a script here? @bdarcus

**Comment by bdarcus** [comment]
I think you can use patterns to validate, but then you lose auto-completion
in validating editors.

>

**Comment by denismaier** [comment]
> I think you can use patterns to validate, but then you lose auto-completion in validating editors.
> […](#)

Ok, so we should have all usable variables in the rnc schema.
But what about the json? Would we get away with a note in the specs?

**Comment by bwiernik** [comment]
I don't really see a problem with listing the variables out in both the rnc and the json schemas if we have the scripts that can generate them automatically when a new variable is added. Put all of the "reviewed", "short", and "original" values together and label as such.

**Comment by denismaier** [comment]
Ok. I can continue working on the script tomorrow, but perhaps just add the variables manually so we can close thus one.
The question is just: -short on everything except? Numbers, names? What about DOI etc? Or should we be liberal here and really allow everything with short, whether it makes sense or not?

And reviewed and original on everything, right?

**Comment by bdarcus** [comment]
No, put them where they obviously belong.

We can always add later.

Variable lists are already getting pretty long.

**Comment by bwiernik** [comment]
Definitely not any names or dates. Other places where they make sense.

**Comment by denismaier** [comment]
Reviewed and original should make sense everywhere right?

**Comment by bwiernik** [comment]
Yes.

**Comment by denismaier** [comment]
Ok. So it's basically short forms for all titles and many strings. Then everything with reviewed and original variants. will do tomorrow.

**Comment by dhimmel** [comment]
Reading through this thread, I'm still not clear on the meaning of `-short`, `original-`, and `reviewed-`? Short makes the most sense to me... styles sometimes want to use a shortened title or container.

But what are original and reviewed? Does this just mean a curator has reviewed the CSL JSON and changed "original" to "reviewed". If this is the case, I don't see a need for these fields to be part of the spec. Users could still include them in CSL JSON. Are styles ever going to render "original" or "reviewed" fields as part of a reference list?

**Comment by bdarcus** [comment]
I think they address cases like these @dhimmel; denis and brenton can correct me if I'm off:

1. a translated title, where you also need the original (untranslated) title
2. a review article, which has its own title, but where one also needs to include the title the article is reviewing (think `review of XYZ`).

**Comment by bwiernik** [comment]
@dhimmel Yes, for example, APA has different formats for reviews of books versus films versus articles. So, at minimum APA would require `reviewed-director`, `reviewed-editor`, and one of `reviewed-type`/`reviewed-genre`/`reviewed-medium`. Other styles additionally require `reviewed-publisher` and `reviewed-issued`. 

For `original-`, there are already `original-` fields for books, but not sufficient fields for other types. For example, if an article is reprinted from another source, APA style wants `original-title` and `original-date` (both already exist), but also `original-container-title`, `original-volume`, `original-issue`, and `original-pages`. 

At this point, the list of `reviewed-` and `original-` variables becomes long enough that we may as well just say "any variable can be supplied with `original-` or `reviewed-` prefixes to refer to the original version of the item or the item being reviewed by the current item.

**Comment by denismaier** [comment]
We might actually also need original type and reviewed type right?

**Comment by bwiernik** [comment]
I think we can avoid it for now.

**Comment by dhimmel** [comment]
Thanks @bwiernik for the explanation of `reviewed-` and `original-`. It seems that they actually both refer to a different work that could have its own CSL Data Item.

It seems to me like the best data model for CSL JSON here would therefore allow the following keys:

```yaml
reviews:
  CSL_Item
original:
  CSL_Item
translation:
  CSL_Item
  language: en-US
```

CSL_Item is a variable here... so it becomes a bit recursive, but gets rid of the repetition.

Sounds like there's a desire to keep the CSL JSON spec as flat as possible. But other fields like date-parts already violate this design. @bwiernik, what are the downsides to the model where fields like `reviews` and `original` point to CSL Item objects themselves?

**Comment by denismaier** [comment]
To be clear, that's something I'd be very much in favour of! However, this is a very major change. Don't know popular the idea is here? Also, I have the impression that some implimentors have concerns about this.

**Comment by bwiernik** [comment]
A major change to the data structure isn't on the table for 1.0.2 or 1.1. I would say that's a CSL 2.0 change.

**Comment by bdarcus** [comment]
We can design now in such a way that a more structured, or relational,
model is an easy step, at least for us.

>

**Comment by bdarcus** [comment]
Like, what if we designed the YAML more structured, with clear logic that
turns that content into flat strings, and then converting to a different
data model later is just adding to those functions?

Still thinking about the idea to have some core, and then an extension
mechanism.

**Comment by denismaier** [comment]
A solid and extensible approach would be great.

The advantage of the current script is that you really have a small list of variables in a couple of categories, and the script will infer from the categories what do do, which variants to create. Having to specify that for each variable can be a downside if a more structured approach. But I wouldn't stand in the way...

**Comment by bdarcus** [comment]
Main thing is clear conventions for the strings (and the scope of
this; per my comments on the PR, I think almost 400 lines for the
variable file is a problem). We can always experiment with the logic
to create them later.

**Comment by denismaier** [comment]
Would splitting it up into multiple files remedy the situation?
  
  
   
    bdarcus <notifications@github.com> hat am 20.06.2020 12:31 geschrieben:
   
   
    
   
   
    
   
   Main thing is clear conventions for the strings (and the scope of
   this; per my comments on the PR, I think almost 400 lines for the
   variable file is a problem). We can always experiment with the logic
   to create them later.

**Comment by bdarcus** [comment]
No.

It's not the file length; it's the number of new variables, per Dan's
concern.

>

**Comment by bdarcus** [comment]
So per [my comment](https://github.com/citation-style-language/schema/pull/263#issuecomment-646969678) on the script PR, I think we need to rethink this.

Tripling or much more the number of variables we offer, and maintain for the next decade or two, does not seem practical.

I asked in the comment if we can could strip this down, and defer the rest to later. Looking at biblatex, which I think is always a good test, shows these "orig" variants:

- title
- date
- language
- publisher
- location

Seems to me rough alignment with biblatex is the more practical first step.

Thinking ahead to some breaking release, I think something like this is maybe where we want to end up with the csl syntax, if we really want to ultimately support all this (but then the obvious question is what apps will support all this?):

``` xml
<text variable="title" relation="original" form="sub"/>
```

**Comment by bwiernik** [comment]
Lost a previous comment I was writing. 

Let’s take a step back and remember what the actual need here is. The data structure is frankly a minor concern. 

_Styles_ need to be able render and format `reviewed-` and `original-` variables, and these are not usually the same format as the parent item. So style authors need to be able to write `reviewed-container-title`, etc. and have it be a valid style. 

These variables could either be enumerated in the RNC or recognized by a pattern and then handled by processor instructions. If there is concern about the lists of variables being long in the RNC (even with explanatory comments), then the pattern recognition+instructions approach would be better. 

In any of this discussion, I think it is really important to remember that the human-readable documentation is the spec, not the schema. The schemas are designed for computers to read. 

For the data model, similarly reviewed and original variables could similarly be recognized by pattern. A flat data model has many advantages. It is easier for an application to create key-value pairs than a hierarchical structure, and especially easier for a user to add a key-value pair to an item if their application doesn’t officially support a variable. 

If a future hierarchical data model were adopted, it would be easy for processors to reconstruct a flat data model by unpacking the reviewed object and constructing the individual reviewed- variables from its contents.

**Comment by bdarcus** [comment]
> Let’s take a step back and remember what the actual need here is. The data structure is frankly a minor concern. In any of this discussion, I think it is really important to remember that the human-readable documentation is the spec, not the schema. The schemas are designed for computers to read.

That's true, except when we're talking data, the obvious question is where that data comes from?

Is zotero, mendeley, etc. going to support this?

What happens if I use biblatex for my source data?

> _Styles_ need to be able render and format `reviewed-` and `original-` variables, and these are not usually the same format as the parent item. So style authors need to be able to write `reviewed-container-title`, etc. and have it be a valid style.

I don't have an issue with titles and dates; the obvious ones.

**Comment by denismaier** [comment]
But biblatex also has a related entry mechanism.
  
  
   
    bdarcus <notifications@github.com> hat am 20.06.2020 14:55 geschrieben:
   
   
    
   
   
    
   
   
   So per my comment on the script PR, I think we need to rethink this.
   Tripling or much more the number of variables we offer, and maintain for the next decade or two, does not seem practical.
   I asked in the comment if we can could strip this down, and defer the rest to later. Looking at biblatex, which I think is always a good test, shows these "orig" variants:
   
    title
    date
    language
    publisher
    location
   
   Seems to me rough alignment with biblatex is the more practical first step.
   Thinking ahead to some breaking release, I think something like this is maybe where we want to end up with the csl syntax, if we really want to ultimately support all this (but then the obvious question is what apps will support all this?):
   
    <text variable="title" relation="original" form="sub"/>

**Comment by bdarcus** [comment]
Maybe that suggests we should reconsider that?

**Comment by bwiernik** [comment]
So, for reviewed and original, just looking at examples in APA, MLA, Chicago, it would need
- title 
- author
- editor
- director
- publisher-place 
- genre
- publisher
- date 
- volume
- issue
- pages
- container-title
- DOI
- ISBN

That’s getting long enough to do a programmatic solution.

**Comment by bdarcus** [comment]
I know @dstillman expressed strong support for a completely flat model, but given how complex this is getting, I think we need to reconsider adding at least a list of "related" objects (reviewed and original most obvoiusly); represented in yaml.

``` yaml
title: Some Title
reviewed:
    type: article
    title: Reviewed Title
    container-title: Newsletter Title
```

... as I suggested above, we could couple it with a corresponding csl attribute:

``` xml
<text variable="title" relation="reviewed"/>
```

Couldn't that work in 1.1?

**Comment by bwiernik** [comment]
> Is zotero, mendeley, etc. going to support this?

Zotero is much more likely to support a key-value pair system than a hierarchical one. Both programs have a mechanism for users supplying additional data in fields that aren’t offered (e.g., by entering them in Extra in Zotero). That’s dramatically easier if it’s a key-value pair system.

**Comment by bdarcus** [comment]
By complexity, BTW, I'm thinking also of style editing.

If I'm editing a style with a validating editor, what do I do with a list of 300 variable options?

**Comment by denismaier** [comment]
As all original- variables will have this prefix I honestly don't see the problem here.
  
  
   
    bdarcus <notifications@github.com> hat am 20.06.2020 15:49 geschrieben:
   
   
    
   
   
    
   
   
   By complexity, BTW, I'm thinking also of style editing.
   If I'm editing a style with a validating editor, what do I do with a list of 300 variable options?

**Comment by bwiernik** [comment]
> Short is different, I think.

Yeah, `short-` is mostly a data entry concern because we have `<text form="short"/>`

**Comment by denismaier** [comment]
So we could get rid of that in the end, but we'd still need it in the json?

**Comment by denismaier** [comment]
> So we could get rid of that in the rnc, but we'd still need it in the json?

**Comment by bdarcus** [comment]
I'll submit a PR later, to show what's involved.

"Short" is not a relation; it's a variant form.

Open question whether a dedicated variable, along with sub, main, etc. But I say add derived data to the form list.

**Comment by bwiernik** [comment]
> If I'm editing a style with a validating editor, what do I do with a list of 100 variable options?

I actually think this might speak in favor of validating styles and data by pattern recognition to avoid populating lists with many variables. 

> <text variable="title" related="reviewed"/>

I do rather like this idea. There could similarly be `<if related="reviewed"/>` as a testable condition. We could specify a limited set of relations (e.g., original, reviewed, retracted, corrected).

**Comment by denismaier** [comment]
Yes, that's interesting. We could also use that to specify relations for name variables, like translator of the title vs translator of the container. Firm could accommodate sub and main.
  
  
   
    Brenton M. Wiernik <notifications@github.com> hat am 20.06.2020 17:37 geschrieben:
   
   
    
   
   
    
   
   
   
    If I'm editing a style with a validating editor, what do I do with a list of 100 variable options?
   
   I actually think this might speak in favor of validating styles and data by pattern recognition to avoid populating lists with many variables.
   
    <text variable="title" related="reviewed"/>
   
   I do rather like this idea. There could similarly be <if related="reviewed"/> as a testable condition. We could specify a limited set of relations (e.g., original, reviewed, retracted, corrected).

**Comment by bwiernik** [comment]
How about we add `reviewed-genre` for 1.0.2 and then revisit related items in v1.1 or v1.2?

**Comment by denismaier** [comment]
> How about we add `reviewed-genre` for 1.0.2 and then revisit related items in v1.1 or v1.2?

You mean **only** this variable now and then discuss the rest later?

**Comment by bwiernik** [comment]
Yes

**Comment by denismaier** [comment]
Deferring is probably a good idea given that we don't seem to find a solution at the moment.

Why do you think we should add exactly this variable? You have something particular in mind?

**Comment by bdarcus** [comment]
That's fine by me, given we already have similar variables in 1.0.1.

But I do think we should make a decision on the future while we're focused on this, since it may impact how we document 1.0.2; whether we want to do it for 1.1, 1.2, or not at all, for example.

I should be able to post the PR in the next day or so. It's super easy to define the attribute, a little more difficult to integrate it into the existing patterns. 

But here's a minimal working schema I used to figure it out:

```
start = element style { cs-text+ }

cs-text =
  element text {
    (alt.data.attributes, vars.base)
    | (alt.data.attributes.titles, vars.titles)
  }

alt.data.attributes = related.attribute?, form.attribute?
alt.data.attributes.titles = related.attribute?, form.attribute.titles?

form.attribute = attribute form { form.attribute.values }
form.attribute.titles = attribute form { form.attribute.values.titles }

related.attribute = attribute related { related.attribute.values }
related.attribute.values = "original" | "reviewed"

form.attribute.values = "long" | "short"
# here we have different attribute values added for titles
form.attribute.values.titles = form.attribute.values | "main" | "sub"

vars.base = attribute variable { vars.base.values }
vars.titles = attribute variable { vars.titles.values }

vars.base.values = "genre"
vars.titles.values = "title" | "container-title"
```

... which validates this; so just a single new attribute, and a few new attribute values, available in the right contexts:

``` xml
<style>
  <text variable="genre" form="short"/>
  <text variable="title" form="maim" related="original"/>
</style>
```

**Comment by denismaier** [comment]
@bdarcus that looks really good. We should of course continue with that discussion. 1.1 is probably a good target.

Your proposal deals mostly with the rnc. What are the consequences for the json schema?

**Comment by bdarcus** [comment]
Conceptually, the json schema changes would be easy.

Just define the new "related" keys with values that are objects, with the
same definitions as the overall schema.

The tricky thing in a transition is converting some of the current flat
variables to this.

So two new keys, and removing maybe 5-10 existing string keys.

**Comment by denismaier** [comment]
Ok. 
What does that mean for Zotero and such? How will they have to provide data? Will they still be able to use a flat data model?

**Comment by bdarcus** [comment]
Not sure. Should be easy enough to support such a change, or for processors
to continue to support current schema.

>

**Comment by dhimmel** [comment]
> Conceptually, the json schema changes would be easy. Just define the new "related" keys with values that are objects, with the same definitions as the overall schema.

I like this implementation for the CSL JSON Schema. Another possibility to consider is that the related field(s) in CSL JSON could just point to other CSL Items by their `id` field. As YAML, this would look like:

```yaml
- id: a-book
  type: book
  title: this is a book that gets reviewed
- id: my-reference
  title: this a book review
  reviews: a-book
```

I don't have a preference at this point, but thought this form might be worth considering. I think the most important thing is that related works are represented as full-fledged CSL Items.

**Comment by bdarcus** [comment]
I agree @dhimmel; worth considering that model too.

I also don't have a preference ATM.

**Comment by bdarcus** [comment]
Actually, your suggestion is potentially better, just in the sense that it
could end up in smaller files, if a user was citing both items and related
items, particularly in a longer document.

**Comment by denismaier** [comment]
Just to make sure it doesn't get list: in that model there will be a relation `original-container` right?

**Comment by denismaier** [comment]
another thing to keep in mind: for certain relations users might want to override data from the related entry. Say, you have a container where issued is a data range, the cited entry is in a particular volume from a specific year...

**Comment by dhimmel** [comment]
> another thing to keep in mind: for certain relations users might want to override data from the related entry.

If you go with the related-item approach (where the related CSL Item is defined within its parent), then overriding won't be an issue, since the related CSL Item only applies to the parent CSL Item.

If you go with the [related-id approach](https://github.com/citation-style-language/schema/issues/167#issuecomment-6470581660), then you could imagine supporting overrides of specific fields. But I think at this point, I think you'd lose the benefits of the related-id approach, in terms of simplicity and avoiding having to add new schema fields.

What about having the user define a separate CSL Item if they need to override anything from a related entry? The situation seems pretty rare, so having to define a separate CSL Item with the exact values required for the relation seems okay. Also if overriding is required, it is probably the case that a different work is being cited... so it should be treated as a distinct work in the CSL JSON?

> Say, you have a container where issued is a data range, the cited entry is in a particular volume from a specific year...

Currently, can a user override values from a CSL JSON Item in a citation? Related works are similar to a CSL Item citing another CSL Item. If citations cannot override reference metadata, then citations-in-references probably shouldn't be able to either.

**Comment by georgd** [comment]
One more thing that should be kept in mind is testing for such variable with multiple attributes. It would be necessary then to do something like

```
<if variable="title" form="short" relation="original">
...
</if>
```

**Comment by bdarcus** [comment]
Yes, this is true.

**Comment by denismaier** [comment]
That should  work, shouldn't it?

**Comment by bwiernik** [comment]
I think it would be reasonably straightforward for Zotero and such to adopt a related-id style CSL JSON structure. Zotero, for example, already related items in its internal database model. It just doesn't do anything with these in the citation generation.

**Comment by bwiernik** [comment]
Regarding `reviewed-genre`, it's the major variable that is missing from being able to describe what type of review a `review` item is. Most reviews could be reasonably be handled with this variable and the existing `reviewed-` variables. Better support for non-book `original-` items requires too many individual variables.

**Comment by bwiernik** [comment]
CSLm has the `cs:alternative` structure. (I don't fully understand how this works, but I think I get the gist.)

There is a testable `context` condition, which can be "alternative" (we could have contexts for "reviewed", "original", "retracted", etc.). Then the same labels are used. So, under the condition `<if `context="reviewed"`>`, the `reviewed-` versions of variables would be used.

Here is an example style implementing `cs:alternative`: https://github.com/Juris-M/jm-styles/blob/master/jm-japan-sociological-society.csl#L827

**Comment by denismaier** [comment]
> One more thing that should be kept in mind is testing for such variable with multiple attributes. It would be necessary then to do something like
> 
> ```
> <if variable="title" form="short" relation="original">
> ...
> </if>
> ```

Thinking about this again: For this to work, I think we need to instruct processors to treat this as a single condition. Including a second test here will obviously get messy and break thinks: 

```
<if variable="title" form="short" relation="original" is-numeric="issued">
...
</if>
 ```

I think `@form` and `@relation` will be applied to both tests, which might or might not be appropriate. So that's basically the some problem as with `@match`. So if this is going to come it will make the case for the extended conditions syntax much stronger.

**Comment by denismaier** [comment]
> Regarding `reviewed-genre`, it's the major variable that is missing from being able to describe what type of review a `review` item is. Most reviews could be reasonably be handled with this variable and the existing `reviewed-` variables. Better support for non-book `original-` items requires too many individual variables.

I've added a PR for this: https://github.com/citation-style-language/schema/pull/269

**Comment by denismaier** [comment]
> CSLm has the `cs:alternative` structure. (I don't fully understand how this works, but I think I get the gist.)
> 
> There is a testable `context` condition, which can be "alternative" (we could have contexts for "reviewed", "original", "retracted", etc.). Then the same labels are used. So, under the condition `<if `context="reviewed"`>`, the `reviewed-` versions of variables would be used.
> 
> Here is an example style implementing `cs:alternative`: https://github.com/Juris-M/jm-styles/blob/master/jm-japan-sociological-society.csl#L827

Yeah, that would be a valuable addition for certain kind of references. This can be used to "to add supplementary reference information to a cited item, such as a translation or reprint." See: https://github.com/citation-style-language/schema/issues/250

The idea is that you can have prefix ordinary variables with `alt-`. Some of those are available as normal variables, but you can also use a special syntax to render them.

I think the basic structure is this:

```xml
<alternative prefix="（" suffix="）">
  <alternative-text/>
</alternative>
```

Essentially, `cs:alternative-text` will render all `alt-` prefixed variables with the current `layout`.

Example:

```xml
<bibliography>
  <layout>
    <group delimiter=". ">
      <text variable="title"/>
      <group delimiter=": ">
        <text variable="publisher-place"/>
        <text variable="publisher"/>
      </group>
    </group>
    <alternative prefix="（" suffix="）">
      <alternative-text/>
    </alternative>
  </layout>
</bibliography>
```

With data:

```yaml
title: A title
publisher: X Press
publisher-place: Y Location
alt-title: The former title
alt-publisher: B Press
alt-publisher-place: C Location
```

The `alt-` variables will be rendered with this part of the style fragment above, but the instead of `title` `alt-title` will be used. 

```xml
 <group delimiter=". ">
      <text variable="title"/>
      <group delimiter=": ">
        <text variable="publisher-place"/>
        <text variable="publisher"/>
      </group>
    </group>
```

We could do something similar with `original` and `reviewed`.

**Comment by bdarcus** [comment]
> ... Another possibility to consider is that the related field(s) in CSL JSON could just point to other CSL Items by their `id` field. As YAML, this would look like:
> 
> ```yaml
> - id: a-book
>   type: book
>   title: this is a book that gets reviewed
> - id: my-reference
>   title: this a book review
>   reviews: a-book
> ```

@dhimmel - to go back to this, in json schema, do you know if we can allow either the above (where `reviews` is a reference id, or also the non-normalized form (where `reviews` is an object)?

 ```yaml
 - id: my-reference
   title: this a book review
   reviews:
      type: book
      title: this is a book that gets reviewed     
 ``` 

if yes, it seems there might be value in that approach?

**Comment by denismaier** [comment]
Yes. Having both options would be best. So you can just use another item if that's easiest. But you could also use the other approach if you need to cite a specific volume of a multi-volume work, and you don't want to add an item just for this volume.

**Comment by bwiernik** [comment]
Supporting both systems is the approach we should take. This would open the door to three useful features that CSL currently lacks:
1. Citations of reviews, etc. being discussed here
2. Parallel legal citation ala CSLm (currently that has to be manually formatted the `references` variable)
3. Support for styles like Annual Reviews, where container information is only given once if multiple items share the same container.

**Comment by bdarcus** [comment]
So it seems we should do this if if we can.

I'd suggest for now we just add two relations ("reviews" and "original"), that we can later expand, with a common "related" definition with this choise.

**Comment by bwiernik** [comment]
That sounds pretty good. We might also want to add a `relationship` variable to describe the nature of the relationship in more detail. For example, "original" could be "adapted from", "reprinted from", "translated from". 

Some potential future expansions:
- "retracted" and "corrected" (this could potentially be specified as one type of "related" with `relationship`)
- For legal cases and patents, "legal-references" (`relationship` could indicate the nature of the reference--affirmed, overturned, etc.)
    - (CSLm currently uses a generic "seeAlso" label for this.)
    - In CSL, we currently have the  `references` variable; I suggest we keep that in place. 
- "container" (for my third use above)

**Comment by denismaier** [comment]
Regarding `relationship `: we should then also add appropriate terms.

**Comment by denismaier** [comment]
Other potential relationships: preprint, ungated versions. And the opposite: journal published versions.

**Comment by bdarcus** [comment]
That probably suggests different modeling then:

``` yaml
relation:
   type: whatever
   refid: book-a
```

**Comment by bwiernik** [comment]
@bdarcus I'm not so sure. For example "reprints", "adapts", and "translates" are types of "original" and could share the same styling coding. 

I suppose instead, we could write styles like this:
```
<if related="original adapts reprints">
   <common code for these types/>
</if>
```

That would accomplish the same thing as a `relationship` variable and be easier to implement.

**Comment by fbennett** [comment]
> CSLm has the cs:alternative structure. (I don't fully understand how this works, but I think I get the gist.)

I haven't revisited the style that exercises the CSL-M `cs:alternative` approach, but I *think* the effects it produces can equally well be achieved in much more transparent CSL code with parallel rendering of independent items joined by a relation. CSL-M (now) implements parallel rendering with a few simple generic attributes that have served well in the IndigoBook implementation, both for parallel cites (some with *very* strange requirements), and for procedural histories (i.e. the chain of appeals following from a single case). It was relatively easy to code in the processor, and has proven to be extremely flexible. It's implemented in the latest `citeproc-js` release, so it can be tested against other use cases. The attributes, which all apply to `cs:group`, are fairly new and need to be properly documented, but they are:
* `parallel-first=<varname ...>`: In a series of related items within a citation, suppresses the group in *all but the first item* if the value of every `varname` matches its *prececessor*.
* `parallel-last=<varname ...>`: In a series of related items within a citation, suppresses the group in *all but the last item* if the value of every `varname` matches its *successor*.
* `parallel-last-to-first=<varname ...>`: When encountered on an item in a related series, this forces variables set with `parallel-last` to be treated as if they are set with `parallel-first`. The change in treatment applies to variables set with `parallel-last` throughout the series, but it affects rendering only in this and subsequent items. (In US legal styles, this can be used to suppress redundant trailing parentheticals in a parallel citation. I can provide examples if the meaning there is not clear...)
* `parallel-delimiter-override`: When encountered on an item in a related series, this overrides the `cs:layout` delimiter between items, for the join with the predecessor.

**Comment by denismaier** [comment]
> 3. Support for styles like Annual Reviews, where container information is only given once if multiple items share the same container.

+1
That's a frequent requirement in German speaking law research.
